src/hibernate-resume/hibernate-resume.c

```diff
@@ -3,36 +3,92 @@
 #include <errno.h>
 #include <sys/stat.h>
 
+#include "device-nodes.h"
 #include "devnum-util.h"
+#include "efivars.h"
 #include "initrd-util.h"
 #include "log.h"
 #include "main-func.h"
 #include "parse-util.h"
+#include "path-util.h"
 #include "sleep-util.h"
+#include "string-util.h"
 
 static const char *arg_resume_device = NULL;
 static uint64_t arg_resume_offset = 0; /* in memory pages */
+static const char *arg_resume_device_efi = NULL;
+static uint64_t arg_resume_offset_efi = 0;
+static bool arg_clear_efi = false;
+
+static int parse_and_validate_arguments(int argc, char *argv[]) {
+        int r;
+
+        if (!IN_SET(argc, 2, 3, 5))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "This program expects one, two, or four arguments.");
+
+        arg_resume_device = empty_to_null(argv[1]);
+        if (argc < 4 && !arg_resume_device)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "No resume device specified.");
+
+        if (!isempty(argv[2])) {
+                r = safe_atou64(argv[2], &arg_resume_offset);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to parse resume offset %s: %m", argv[2]);
+        }
+
+        if (argc < 4)
+                return 0;
+
+        arg_resume_device_efi = argv[3];
+
+        r = safe_atou64(argv[4], &arg_resume_offset_efi);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse EFI HibernateLocation resume offset %s: %m", argv[4]);
+
+        arg_clear_efi = true;
+
+        if (!arg_resume_device) {
+                arg_resume_device = arg_resume_device_efi;
+                arg_resume_offset = arg_resume_offset_efi;
+        } else {
+                if (!path_equal(arg_resume_device, arg_resume_device_efi)) {
+                        r = devnode_same(arg_resume_device, arg_resume_device_efi);
+                        if (r < 0)
+                                log_warning_errno(r,
+                                                  "Failed to check if resume=%s is the same device as EFI HibernateLocation device '%s', ignoring: %m",
+                                                  arg_resume_device, arg_resume_device_efi);
+                        if (r == 0)
+                                log_warning("resume=%s doesn't match with EFI HibernateLocation device '%s', proceeding anyway with resume=.",
+                                            arg_resume_device, arg_resume_device_efi);
+                }
+
+                if (arg_resume_offset != arg_resume_offset_efi)
+                        log_warning("resume_offset=%" PRIu64 " doesn't match with EFI HibernateLocation offset %" PRIu64 ", proceeding anyway with resume_offset=.",
+                                    arg_resume_offset, arg_resume_offset_efi);
+        }
```

Comment:
hmm, i think it would be a tiny bit nicer, to move the the efi var parsing into common code somewhere (including the superficial checks that compare the data with our own /etc/os-release), and then parse it once from the generator and do the superficial checks, and then parse it also here, and do also the thorough checks.

i think this would make this a bit more robust, since the generator would be a bit more optional, and we'd always go to the full source of truth and validate everything.

And yes, I am aware this would make things asymmetric: the kernel cmdline option would be parsed in the generator and passed to the tool via its cmdline, and the efi var would be read by both directly. But I think that's OK, since the kernel cmdline needs less verification, as it doesn't store os info data and suchlike.